File: solutions/standing-ovation/go/standing-ovation.go
package main

import (
	"fmt"
	"io/ioutil"
	"os"
	"strconv"
	"strings"
)

func main() {
	pwd, _ := os.Getwd()
	file, err := ioutil.ReadFile(pwd + "/" + os.Args[1])

	if err != nil {
		fmt.Println(err)
		return
	}

	input := string(file)
	input = input[:len(input)-1]
	results := getResults(input)
	printResults(results)
}

func printResults(results []int) {
	for i := 0; i < len(results); i++ {
		fmt.Printf("Case #%v: %v\n", i+1, results[i])
	}
}

func getResults(input string) []int {
	lines := strings.Split(input, "\n")
	var results []int

	for i := 1; i < len(lines); i++ {
		testCase := lines[i]
		results = append(results, getTestCaseResult(testCase))
	}

	return results
}

func getTestCaseResult(testCase string) int {
	maxShyness, _ := strconv.Atoi(strings.Split(testCase, " ")[0])
	audience := strings.Split(strings.Split(testCase, " ")[1], "")

	standingPeopleCount := 0
	friendsCount := 0

	for shynessLevel := 0; shynessLevel <= maxShyness; shynessLevel++ {
		peopleCount, _ := strconv.Atoi(audience[shynessLevel])

		if shynessLevel > standingPeopleCount {
			friendsNeeded := shynessLevel - standingPeopleCount
			friendsCount += friendsNeeded
			standingPeopleCount += friendsNeeded + peopleCount
		} else {
			standingPeopleCount += peopleCount
		}
	}

	return friendsCount
}

File: solutions/standing-ovation/ruby/standingOvation.rb
def getResults(args)
    lines = args.split("\n")
    results = []

    return "Error" unless isTestCasesQuantityValid(lines)

    for i in 1...lines.length
      line = lines[i]
      return "Error" unless isTestCaseValid(line)

      results.push(getTestCaseResult(line))
    end

    results
end

def isTestCasesQuantityValid(lines)
  testCasesQuantity = Integer(lines[0])
  testCasesQuantity >= 1 && testCasesQuantity <= 100 && lines.length - 1 == testCasesQuantity
end

def isTestCaseValid(testCase)
  testCaseLine = testCase.split(" ")
  maxShyness = Integer(testCaseLine[0])
  audienceString = testCaseLine[1]

  maxShyness == audienceString.length - 1 && maxShyness >= 0 && maxShyness <= 1000
end

def getTestCaseResult(testCase)
        maxShyness = Integer(testCase.split(" ")[0])
        audience = testCase.split(" ")[1].split("")

        standingPeopleCount = 0
        friendsCount = 0

        for shynessLevel in 0..maxShyness
            peopleCount = Integer(audience[shynessLevel])
            if shynessLevel > standingPeopleCount
                friendsNeeded = shynessLevel - standingPeopleCount
                friendsCount += friendsNeeded
                standingPeopleCount += friendsNeeded + peopleCount
            else
                standingPeopleCount += peopleCount
            end
        end

        friendsCount
end

file = File.open("./A-small-practice.in", "r")
contents = file.read

results = getResults(contents)
results.each_with_index do |result, i|
  puts "Case ##{i + 1}: #{result}"
end
